### PR TITLE
bpo-46581: Propagate private vars via _GenericAlias.copy_with

### PR DESCRIPTION
**Pull request by posita**
GH-26091 added the `_typevar_types` and `_paramspec_tvars` instance variables to `_GenericAlias`. However, they were not propagated consistently. This commit addresses the most prominent deficiency identified in [bpo-46581](https://bugs.python.org/issue46581) (namely their absence from `_GenericAlias.copy_with`), but there could be others.

https://bugs.python.org/issue46581

### CODE REVIEW

File: Lib/typing.py [hunk 1]
```diff
@@ -325,8 +325,8 @@ def inner(*args, **kwds):
 def _eval_type(t, globalns, localns, recursive_guard=frozenset()):
     """Evaluate all forward references in the given type t.
     For use of globalns and localns see the docstring for get_type_hints().
-    recursive_guard is used to prevent prevent infinite recursion
-    with recursive ForwardRef.
+    recursive_guard is used to prevent infinite recursion with a recursive
+    ForwardRef.
```
Comment (posita) [inline Lib/typing.py]: I found this while researching bpo-46581. This edit removes a duplicate "prevent" and re-justifies text. I'm happy to remove this edit if it was intended as a subtle recursion prevention joke that was lost on me.

File: Lib/typing.py [hunk 2]
```diff
@@ -964,7 +964,7 @@ def __getattr__(self, attr):
             return self._name or self.__origin__.__name__
 
         # We are careful for copy and pickle.
-        # Also for simplicity we just don't relay all dunder names
+        # Also for simplicity we don't relay any dunder names
```
Comment (posita) [inline Lib/typing.py]: This comment was something I tripped over while researching bpo-46581. I don't understand what the original means. I think the above captures the sentiment more clearly. I'm happy to remove this edit, if I missed the mark.

File: Lib/test/test_genericalias.py [hunk 3]
```diff
@@ -298,6 +298,14 @@ def __deepcopy__(self, memo):
             self.assertEqual(copied.__args__, alias.__args__)
             self.assertEqual(copied.__parameters__, alias.__parameters__)
 
+    def test_copy_with(self):
```
Comment (JelleZijlstra) [inline Lib/test/test_genericalias.py]: Shouldn't this be in test_typing, since it tests the typing.py implementation? This file should test `types.GenericAlias`.

Also `copy_with` is an undocumented implementation detail. Is there an example of something that a user of typing.py would do that is fixed by this change?

Comment (posita) [reply Lib/test/test_genericalias.py]: I'm happy to move it to `test_typing` if you think that's where it should live. I was guessing about `test_genericalias`.

Regarding your question about `copy_with`, see [bpo46581](https://bugs.python.org/issue46581) for details, especially Ken Jin's detailed explanation. He writes (in pertinent part):

> ```
> I can reproduce a semi-bug right now:
> 
> >>> P = ParamSpec('P')
> >>> Callable[P, int].__parameters__
> (~P,)
> >>> Callable[P, int].copy_with((P,int))
> typing.Callable[~P, int]
> >>> Callable[P, int].copy_with((P,int)).__parameters__
> ()
> ^ This shouldn't be empty!!
> 
> Would you like to submit a patch for this?
> ```

The issue was filed at [GVR's request](https://github.com/python/cpython/pull/26091#issuecomment-1024900261) after my prodding. That and Ken Jin's invitation to submit a patch resulted in this PR.

That's not directly responsive, but it's about as close as I can come to a real answer.

Comment (posita) [reply Lib/test/test_genericalias.py]: Apologies for the stealth edit. I added some additional color above. ☝️

Comment (posita) [reply Lib/test/test_genericalias.py]: Hmmm…it looks like this has suffered from bit rot, which is unfortunate. Tests no longer pass as of ecfacc3 (#27518, cc @serhiy-storchaka). I now get the following errors:

``` py
======================================================================
ERROR: test_concatenate (test.test_typing.CollectionsCallableTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/…/cpython/Lib/test/test_typing.py", line 757, in test_concatenate
    self.assertEqual(C[P2, int], Callable[Concatenate[int, P2], int])
                     ~^^^^^^^^^
  File "<frozen _collections_abc>", line 502, in __getitem__
  File "/…/cpython/Lib/typing.py", line 321, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1160, in __getitem__
    r = self.copy_with(new_args)
        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1436, in copy_with
    return super().copy_with(params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1219, in copy_with
    return self.__class__(self.__origin__, args, name=self._name, inst=self._inst,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1424, in __init__
    super().__init__(*args, **kwargs,
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: typing._GenericAlias.__init__() got multiple values for keyword argument '_typevar_types'

======================================================================
ERROR: test_var_substitution (test.test_typing.ConcatenateTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/…/cpython/Lib/test/test_typing.py", line 5331, in test_var_substitution
    self.assertEqual(C[int, P2], Concatenate[int, P2])
                     ~^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 321, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1160, in __getitem__
    r = self.copy_with(new_args)
        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1436, in copy_with
    return super().copy_with(params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1219, in copy_with
    return self.__class__(self.__origin__, args, name=self._name, inst=self._inst,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1424, in __init__
    super().__init__(*args, **kwargs,
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: typing._GenericAlias.__init__() got multiple values for keyword argument '_typevar_types'

======================================================================
ERROR: test_concatenate (test.test_typing.TypingCallableTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/…/cpython/Lib/test/test_typing.py", line 757, in test_concatenate
    self.assertEqual(C[P2, int], Callable[Concatenate[int, P2], int])
                     ~^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 321, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1159, in __getitem__
    new_args = self._determine_new_args(args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1197, in _determine_new_args
    new_arg = old_arg[subargs]
              ~~~~~~~^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 321, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1160, in __getitem__
    r = self.copy_with(new_args)
        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1436, in copy_with
    return super().copy_with(params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1219, in copy_with
    return self.__class__(self.__origin__, args, name=self._name, inst=self._inst,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/…/cpython/Lib/typing.py", line 1424, in __init__
    super().__init__(*args, **kwargs,
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: typing._GenericAlias.__init__() got multiple values for keyword argument '_typevar_types'

```

If I back out my change to `copy_with`, my own test still shows a propagation problem:

``` py
======================================================================
FAIL: test_copy_with (test.test_typing.GenericTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/…/cpython/Lib/test/test_typing.py", line 2341, in test_copy_with
    self.assertEqual(original.__parameters__, copied.__parameters__)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Tuples differ: (~P,) != ()

First tuple contains 1 additional elements.
First extra element 0:
~P

- (~P,)
+ ()

```

So, as you point out @JelleZijlstra, it seems like `copy_with` is a(n unfortunately named and) fairly obscure internal implementation detail. Not sure how to proceed here?

File: Lib/typing.py [hunk 4]
```diff
@@ -1216,7 +1216,9 @@ def _determine_new_args(self, args):
         return tuple(new_args)
 
     def copy_with(self, args):
-        return self.__class__(self.__origin__, args, name=self._name, inst=self._inst)
+        return self.__class__(self.__origin__, args, name=self._name, inst=self._inst,
+                              _typevar_types=self._typevar_types,
+                              _paramspec_tvars=self._paramspec_tvars)
```
Comment (posita) [inline Lib/typing.py:1346]: This (perplexingly) no longer works after merging #27518.

File: Lib/typing.py (hunk 2 above)
Comment (Fidget-Spinner) [reply Lib/typing.py]: Sorry, do you mind splitting out the copyedits into another PR please? I plan to backport them to 3.9 and that will make that easier. You don't have to create a new bpo issue just for that, just a normal PR will do. Thank you!

Comment (posita) [reply Lib/typing.py]: Done: #31684. (And removed from this PR.)

File: Lib/typing.py (hunk 4 above)
Comment (Fidget-Spinner) [reply Lib/typing.py:1346]: Do your tests pass with https://github.com/python/cpython/pull/27518 but without modifying this line of code? I have a suspicion that PR accidentally fixed our bug too.

If that's the case, we just need to merge the tests here.

Comment (posita) [reply Lib/typing.py:1346]: Sorry for being vague. #27518 breaks this PR.

After #27518, and **_with_** the change highlighted by this comment thread, these tests now fail:

* `ERROR: test_concatenate (test.test_typing.CollectionsCallableTests)`
* `ERROR: test_var_substitution (test.test_typing.ConcatenateTests)`
* `ERROR: test_concatenate (test.test_typing.TypingCallableTests)`

After #27518, but **_without_** this change, the test introduced elsewhere in this PR fails:

* `FAIL: test_copy_with (test.test_typing.GenericTests)`

So #27518 breaks my approach for `_GenericAlias.copy_with`, but the problem identified in [bpo-46581](https://bugs.python.org/issue46581) still remains. I'm still trying to grok why, but `typing.py` internals are very hard to follow, with a lot of non-intuitive interactions and layers that seem to grow as time goes by, not recede.

I may be naive, but, intuitively, the change herein _should_ make sense. #26091 added members to the class and exposed setting them via the `__init__` method. The `copy_with` method already invoked the `__init__` method to make a copy. This change propagates the new members via the `__init__` method, which arguably should have been done as part of #26091.

#27518 seems to add even more layers of indirection/complexity. It's not even obvious how those changes interact with these. Perhaps it just isn't for me to understand. Maybe there's genius at work here, but some overarching design (if it exists) is hard for me to tease out. To me, it just appears as a nearly unnavigable hodgepodge of intertwined kludges. 🤷

I think at this point, this will likely require someone more familiar with that code base. My inclination is to close this PR as a failed attempt and leave [bpo-46581](https://bugs.python.org/issue46581) open for someone else to fix, since I'm unclear how to proceed in light of #27518.

File: Lib/typing.py [hunk 5]
```diff
@@ -1217,7 +1217,9 @@ def _determine_new_args(self, args):
         return tuple(new_args)
 
     def copy_with(self, args):
-        return self.__class__(self.__origin__, args, name=self._name, inst=self._inst)
+        return self.__class__(self.__origin__, args, name=self._name, inst=self._inst,
```
Comment (Fidget-Spinner) [inline Lib/typing.py:1344]: I found the issue. The problem is that `_ConcatenateGenericAlias` inherits from `_GenericAlias`. That PR calls `super().copy_with` and passes everything up. Now if we look at `_ConcatenateGenericAlias`'s `__init__`:
https://github.com/python/cpython/blob/9510c9511a3596291523b671f8bc59f581e8f2fb/Lib/typing.py#L1425-L1427

It effectively expands to:
```python
 super().__init__(*args, _typevar_types=self._typevar_types, _paramspec_tvars=self._paramspec_tvars, 
                  _typevar_types=(TypeVar, ParamSpec), 
                  _paramspec_tvars=True) 
```
Note the duplicates! That's why the error message says `TypeError: typing._GenericAlias.__init__() got multiple values for keyword argument '_typevar_types'`. The solution is to special-case `_ConcatenateGenericAlias`.
```suggestion
        if isinstance(self, _ConcatenateGenericAlias):
            return self.__class__(self.__origin__, args)
        return self.__class__(self.__origin__, args, name=self._name, inst=self._inst,
```

Comment (posita) [reply Lib/typing.py:1344]: Thanks for the investigation and for the proposal! This is actually quite depressing, though. I'm really sad if `_GenericAlias` now has to be aware of one of its child classes `_ConcatenateGenericAlias`, and carve out an exception in the base implementation. Doesn't that seem backwards? It may be the only practical option we have given the current state of things, but isn't this like a rookie OOD smell? Doesn't this further contribute to the spaghettification of `typing.py`?

Comment (posita) [reply Lib/typing.py:1344]: Actually, I take that back. This might be unavoidable? I think the "design" of `copy_with` itself is broken. It looks like it's trying to pull double- or triple-duty depending on who's using it and where. In that world, all OOD bets are off, so this might be making the best of a bad situation?

Or maybe `_ConcatenateGenericAlias.__init__(…, **kwargs, _typevar_types=(TypeVar, ParamSpec), _paramspec_tvars=True)` is the wrong way to go? Not sure. This seems like one hack on top of another. What's one more? 😬

File: Misc/NEWS.d/next/Library/2022-02-01-11-32-47.bpo-46581.t7Zw65.rst [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Propagates :attribute:`~typing._GenericAlias._typevar_types` and
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2022-02-01-11-32-47.bpo-46581.t7Zw65.rst]: All this are implementation details. A common user knows nothing about it. Describe how the bug affected a common user.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
There is more straightforward way to fix a conflict.
1. Add `_typevar_types=(TypeVar, ParamSpec), _paramspec_tvars=True` to the call of `_ConcatenateGenericAlias` in `_Concatenate`.
2. Remove `_ConcatenateGenericAlias.__initt__`.
3. Remove a special case from `_GenericAlias.copy_with`.

Also please move changes in comments and docstrings to enother PR. They distract.

File: Misc/NEWS.d/next/Library/2022-02-01-11-32-47.bpo-46581.t7Zw65.rst (hunk 6 above)
Comment (posita) [reply Misc/NEWS.d/next/Library/2022-02-01-11-32-47.bpo-46581.t7Zw65.rst]: Updated: https://github.com/python/cpython/pull/31061/files#diff-342ce6eb72310ee05e41d8897f6726904fd2bcbe83c79cf4f58e344e366db1c4.

**Comment by posita** [comment]
> There is more straightforward way to fix a conflict.
> 
> 1. Add `_typevar_types=(TypeVar, ParamSpec), _paramspec_tvars=True` to the call of `_ConcatenateGenericAlias` in `_Concatenate`.
> 2. Remove `_ConcatenateGenericAlias.__initt__`.
> 3. Remove a special case from `_GenericAlias.copy_with`.
> 
> Also please move changes in comments and docstrings to enother PR. They distract.

Nice! 🙏❤️ Done.

**Comment by JelleZijlstra** [comment]
It's still not clear to me what user-visible bug this actually fixes. `.copy_with()` is an undocumented implementation detail. Can you provide a test case fixed by this PR that doesn't call `copy_with` directly?

**Comment by posita** [comment]
> It's still not clear to me what user-visible bug this actually fixes. `.copy_with()` is an undocumented implementation detail. Can you provide a test case fixed by this PR that doesn't call `copy_with` directly?

Other than @Fidget-Spinner's example, I don't have one.

@gvanrossum, do you want to weigh in on this? This PR attempts to fix the [the issue filed at your request](https://github.com/python/cpython/pull/26091#issuecomment-1024900261).

**Comment by gvanrossum** [comment]
> > It's still not clear to me what user-visible bug this actually fixes. `.copy_with()` is an undocumented implementation detail. Can you provide a test case fixed by this PR that doesn't call `copy_with` directly?
> 
> Other than @Fidget-Spinner's example, I don't have one.
> 
> @gvanrossum, do you want to weigh in on this? This PR attempts to fix the [the issue filed at your request](https://github.com/python/cpython/pull/26091#issuecomment-1024900261).

Sorry, I haven't followed this in enough detail. I thought that you were pointing out in that comment that something was amiss with the previous PR. Could you produce a unit test based on @Fidget-Spinner's example? Or is that not the right way to think about it?

File: Lib/test/test_typing.py [hunk 7]
```diff
@@ -2802,6 +2802,14 @@ def __init__(self, attr: T) -> None:
...
         self.assertEqual(ci.__orig_class__, C[int])
 
+    def test_copy_with(self):
+        # bpo-46581
+        from typing import Callable, ParamSpec
+        P = ParamSpec('P')
+        original = Callable[P, int]
+        copied = original.copy_with((P, int))
+        self.assertEqual(original.__parameters__, copied.__parameters__)
+
```
Comment (posita) [inline Lib/test/test_typing.py]: ☝️ @gvanrossum

Comment (posita) [reply Lib/test/test_typing.py:2812]: Modeled after [this](https://bugs.python.org/msg412149).

**Comment by posita** [comment]
> Sorry, I haven't followed this in enough detail. I thought that you were pointing out in that comment that something was amiss with the previous PR. Could you produce a unit test based on @Fidget-Spinner's example? Or is that not the right way to think about it?

There is already one as part of this PR modeled after @Fidget-Spinner's [original observation](https://bugs.python.org/msg412149).

File: Lib/test/test_typing.py [hunk 8]
```diff
@@ -2802,6 +2802,14 @@ def __init__(self, attr: T) -> None:
         self.assertEqual(deepcopy(ci).attr, 1)
         self.assertEqual(ci.__orig_class__, C[int])
 
+    def test_copy_with(self):
+        # bpo-46581
+        from typing import Callable, ParamSpec
```
Comment (serhiy-storchaka) [inline Lib/test/test_typing.py]: Aren't they imported at the top of the file?

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

File: Lib/test/test_typing.py (hunk 8 above)
Comment (posita) [reply Lib/test/test_typing.py]: Yeah, fair. Removed. (These were residual from originally placing this test in `test_genericalias.py`.)

**Comment by posita** [comment]
> That test relies on calling `.copy_with()` directly, which is not a public API.
> 
> …
> 
> The last line is wrong and I think this PR fixes it.

Thanks @JelleZijlstra! Apologies for all the back-and-forth. The [latest modification to `Lib/test/test_typing.py`](https://github.com/python/cpython/pull/31061/files#diff-04d29c98076c2d6bb75921ea9becb26a862544d39b71db87b6e354c759b9305d) should address that.

File: Lib/typing.py [hunk 9]
```diff
@@ -1339,7 +1341,9 @@ def _determine_new_args(self, args):
...
+                              _typevar_types=self._typevar_types,
+                              _paramspec_tvars=self._paramspec_tvars)
```
Comment (posita) [inline Lib/typing.py:1346]: The newly-added test fails without this addition.

File: Lib/test/test_typing.py [hunk 10]
```diff
@@ -2802,6 +2802,14 @@ def __init__(self, attr: T) -> None:
         self.assertEqual(deepcopy(ci).attr, 1)
         self.assertEqual(ci.__orig_class__, C[int])
 
+    def test_parameter_propagation(self):
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: I think this would fit better in a test class for ParamSpec. Also, it's worth testing that `__parameters__` is still right in more complicated cases, e.g. if the P is inside a Concatenate.

Comment (posita) [reply Lib/test/test_typing.py]: I don't think one exists right now. I'm happy to create one. (`ParamSpecTestCase`?) Should that belong in its own file, or a separate class inside `test_typing.py`?

Comment (JelleZijlstra) [reply Lib/test/test_typing.py]: There is a `ParamSpecTests` class.

Comment (posita) [reply Lib/test/test_typing.py]: Ah! Duh. I'm an idiot.

Moved.

File: Lib/test/test_typing.py [hunk 11]
```diff
@@ -5793,6 +5793,14 @@ def test_paramspec_in_nested_generics(self):
         self.assertEqual(G2[[int, str], float], list[C])
         self.assertEqual(G3[[int, str], float], list[C] | int)
 
+    def test_paramspec_gets_copied(self):
+        # bpo-46581
+        P = ParamSpec('P')
+        original = Callable[P, int]
+        self.assertEqual(original.__parameters__, (P,))
+        copied = original[P]
+        self.assertEqual(original.__parameters__, copied.__parameters__)
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: Could you add some more complicated test cases? Ideas:
- The original has a `Concatenate[int, P]` in it (`original = Callable[Concatenate[int, P], int]; original[P].__paramaters__ == (P,)`)
- Instead of substituting a `P` in, substitute in a `Concatenate` or something like `[int, T]`, and then the `T` should be in `__parameters__`.

Comment (serhiy-storchaka) [inline Lib/test/test_typing.py]:
```suggestion
        P = ParamSpec('P')
        P2 = ParamSpec('P2')
        C1 = Callable[P, int]
        self.assertEqual(C1.__parameters__, (P,))
        self.assertEqual(C1[P2].__parameters__, (P2,))
        self.assertEqual(C1[str].__parameters__, ())
        self.assertEqual(C1[str, T].__parameters__, (T,))
        self.assertEqual(C1[Concatenate[str, P2]].__parameters__, (P2,))
        self.assertEqual(C1[Concatenate[T, P2]].__parameters__, (T, P2))
        self.assertEqual(C1[...].__parameters__, ())

        C2 = Callable[Concatenate[str, P], int]
        self.assertEqual(C2.__parameters__, (P,))
        self.assertEqual(C2[P2].__parameters__, (P2,))
        self.assertEqual(C2[str].__parameters__, ())
        self.assertEqual(C2[str, T].__parameters__, (T,))
        self.assertEqual(C2[Concatenate[str, P2]].__parameters__, (P2,))
        self.assertEqual(C2[Concatenate[T, P2]].__parameters__, (T, P2))
```

File: Lib/test/test_typing.py [hunk 12]
```diff
@@ -5793,6 +5793,27 @@ def test_paramspec_in_nested_generics(self):
         self.assertEqual(G2[[int, str], float], list[C])
         self.assertEqual(G3[[int, str], float], list[C] | int)
 
+    def test_paramspec_gets_copied(self):
+        # bpo-46581
+        P = ParamSpec('P')
+        P2 = ParamSpec('P2')
+        C1 = Callable[P, int]
+        self.assertEqual(C1.__parameters__, (P,))
+        self.assertEqual(C1[P2].__parameters__, (P2,))
```
Comment (serhiy-storchaka) [inline Lib/test/test_typing.py:5802]: Only this test was failed.

**Review by serhiy-storchaka — APPROVED** [review]

**Review by Fidget-Spinner — APPROVED** [review]
Thank you for pulling through @posita . This was a pretty tricky PR.

**Comment by posita** [comment]
> Sorry, @posita and @serhiy-storchaka, I could not cleanly backport this to `3.10` due to a conflict. Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line. `cherry_picker 32bf3597922ac3f613989582afa2bff43bea8a2f 3.10`

Should I work on resolving this?

**Comment by posita** [comment]
I would love a nudge on where to learn how to do this manually. (I'm unfamiliar with the process.) Is it as simple as resolving conflicts and then submitting a PR against the [3.10 branch](https://github.com/python/cpython/tree/3.10) or is there more involved?

**Comment by Fidget-Spinner** [comment]
> I would love a nudge on where to learn how to do this manually. (I'm unfamiliar with the process.) Is it as simple as resolving any conflicts against the [3.10 branch](https://github.com/python/cpython/tree/3.10) and then submitting a PR or is there more involved?

Yup. Submit a PR against the 3.10 and/or 3.9 branch after resolving conflicts.

The bot is a little off. An easier way without installing things is to use the `git cherry-pick` command directly. StackOverflow link [here](https://stackoverflow.com/questions/9339429/what-does-cherry-picking-a-commit-with-git-mean) does a great job explaining it.

The other even easier alternative if you're using GitHub desktop, is to drag and drop the commit from the left panel from main into your new branch that is based off 3.10/3.9. The app will then warn you about conflicts, and you can resolve them in a code editor (this means accepting/rejecting changes or overwriting them altogether manually).

**Comment by posita** [comment]
> > I would love a nudge on where to learn how to do this manually. (I'm unfamiliar with the process.) Is it as simple as resolving any conflicts against the [3.10 branch](https://github.com/python/cpython/tree/3.10) and then submitting a PR or is there more involved?
> 
> Yup. Submit a PR against the 3.10 and/or 3.9 branch after resolving conflicts.
> 
> …

Thanks @Fidget-Spinner! Cherry-picking is all good. I just didn't know if there was other repo- or CI-specific magical incantations I needed to invoke or procedures I needed to follow.

Question (maybe for @JelleZijlstra): This is marked as needed a back-port to 3.10, but not to 3.9. Should this also be back-ported to 3.9?

**Comment by Fidget-Spinner** [comment]
@posita re backporting to 3.9, we don't need it as pre-3.10 ParamSpec didn't exist so we never needed these parameters then.

Anyways, congrats on becoming a CPython contributor!

**Comment by posita** [comment]
> @posita re backporting to 3.9, we don't need it as pre-3.10 ParamSpec didn't exist so we never needed these parameters then.

Oh, right! Duh. Thanks all! I very much appreciate all the hand-holding.
